### PR TITLE
fix: add cron job to pick up hanging prs

### PR DESCRIPTION
**Pull request by sofisl**
This addresses #1159 

Following Jeff's suggestion, this PR adds functionality for a cron job that will run once an hour. 

I *believe* that, if we format the cron job correctly, we will be able to trigger the gcf-utils mechanism of rotating through all of the repositories under which MOG is installed. 

Taking a lead from the SRS cron job, I think we can just check if the `cron_org` exists in the payload to trigger the rotation through all of the repos, which is different from the current MOG cron job in Cloud Scheduler.

I am cautious about this approach - we may trigger an abuse mechanism or hit our rate limit faster since we'd be checking all PRs in all our repos. I feel like we should try this solution, but if it doesn't work, we may want to write another bot that just picks up PRs.

### CODE REVIEW

File: packages/merge-on-green/src/merge-on-green.ts [hunk 1]
```diff
@@ -331,6 +331,98 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+handler.pickUpPR = async function pickUpPR(
+  owner: string,
+  repo: string,
+  github: Context['github']
+) {
+  const prs = await github.paginate(await github.pulls.list, {
+    owner,
+    repo,
+    state: 'open',
+  });
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:350]: You can probably reduce the # of queries, by using the search API: https://docs.github.com/en/free-pro-team@latest/rest/reference/search#search-issues-and-pull-requests

File: packages/merge-on-green/src/merge-on-green.ts [hunk 2]
```diff
@@ -331,6 +331,98 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+      } catch (err) {
+        err.message = `Error in getting branch protection\n\n${err.message}`;
+        await github.issues.createComment({
+          owner,
+          repo,
+          issue_number: pr.number,
+          body:
+            "Your PR doesn't have any required checks. Please add required checks to your master branch and then re-add the label. Learn more about enabling these checks here: https://help.github.com/en/github/administering-a-repository/enabling-required-status-checks.",
+        });
+        logger.error(err);
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:391]: Should this logic share the same code as the webhook handling?

**Review by chingor13 — CHANGES_REQUESTED** [review]

File: packages/merge-on-green/src/merge-on-green.ts [hunk 3]
```diff
@@ -331,6 +331,98 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+      }
+    );
+
+    const labelFound = labels.find(
+      (label: Label) =>
+        label.name === MERGE_ON_GREEN_LABEL ||
+        label.name === MERGE_ON_GREEN_LABEL_SECURE
+    );
+
+    if (labelFound) {
```
Comment (bcoe) [inline packages/merge-on-green/src/merge-on-green.ts:368]: we should only do this if there's not already an entry in `datastore` for the given PR right?

File: packages/merge-on-green/src/merge-on-green.ts (hunk 2 above)
Comment (bcoe) [reply packages/merge-on-green/src/merge-on-green.ts:391]: 👍 I agree, I think the logic could be simplified if we:

1. check to see if the PR is already tracked in datastore (_if it is tracked do nothing._).
2. if the PR is not tracked, but an automerge label exists, make a synthetic call to `app.on('pull_request.labeled', async context`; I think you could do this by looking at the fields you use from context, and then making sure you populate these in a fake context object you pass.

**Review by bcoe — CHANGES_REQUESTED** [review]
Added a couple notes. I agree with @chingor13, I think you could be clever and make it so that finding a labeled PR that isn't tracked hits the same code path as adding a label.

File: packages/merge-on-green/src/merge-on-green.ts (hunk 2 above)
Comment (chingor13) [reply packages/merge-on-green/src/merge-on-green.ts:391]: I'm not sure about faking an webhook - we could probably extract that common logic into a function to execute from both paths.

Comment (sofisl) [reply packages/merge-on-green/src/merge-on-green.ts]: I simplified the logic here!

**Comment by sofisl** [comment]
@chingor13 and @bcoe, made some adjustments, thanks for the review!

File: packages/merge-on-green/src/merge-on-green.ts [hunk 4]
```diff
@@ -331,6 +432,77 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
 
+//TODO: change the search query if expanding the allowlist to search for more authors
+/**
+ * For a given repository, looks through all the PRs and checks to see if they have a MOG label
+ * @param owner the owner of the repo
+ * @param repo the repo name
+ * @param context the context of the webhook payload
+ * @returns void
+ */
+handler.pickUpPR = async function pickUpPR(
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:443]: nit: can we call this `scanForMissingPullRequests` or something like that?

File: packages/merge-on-green/src/merge-on-green.ts [hunk 5]
```diff
@@ -214,29 +288,56 @@ handler.checkIfPRIsInvalid = async function checkIfPRIsInvalid(
...
-      reactionId: wp.reactionId,
-      installationId: wp.installationId,
-    },
-    method: 'upsert',
-  };
-  await datastore.save(entity);
+handler.addPR = async function addPR(
+  wp: WatchPR,
+  url: string,
+  github: Context['github']
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:300]: Other bots are using this type: https://github.com/googleapis/repo-automation-bots/blob/78950c856ed7205341efd8908b4cca935ab1c5ab/packages/gcf-utils/src/gcf-utils.ts#L41

File: packages/merge-on-green/src/merge-on-green.ts [hunk 6]
```diff
@@ -39,12 +39,12 @@ interface WatchPR {
...
   owner: string;
   state: 'continue' | 'stop';
-  branchProtection: string[];
+  branchProtection?: string[];
   label: string;
   author: string;
   url: string;
-  reactionId: number;
-  installationId: number;
+  reactionId?: number | undefined;
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:46]: nit: is `undefined` implicitly possible with the `?`

File: packages/merge-on-green/src/merge-on-green.ts [hunk 7]
```diff
@@ -350,6 +522,22 @@ function handler(app: Application) {
...
+      if (context.payload.cron_org !== owner) {
+        logger.info(`skipping run for ${context.payload.cron_org}`);
+        return;
+      }
+      await handler.pickUpPR(owner, repo, context.github);
+      return;
+    }
+
     const start = Date.now();
     await handler.checkPRMergeability(watchedPRs, app, context);
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:542]: Is it worth trying to do these both async and awaiting them at the end? (it's going to complete all of `pickUpPR` before trying to do the merges.)

File: packages/merge-on-green/src/merge-on-green.ts [hunk 8]
```diff
@@ -331,6 +432,77 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+  repo: string,
+  github: Context['github']
+) {
+  // Github does not support searching the labels with 'OR'.
+  // The searching for issues is considered to be an "AND" instead of an "OR" .
+  const issuesAutomergeLabel = await github.paginate(
+    await github.search.issuesAndPullRequests,
+    {
+      q:
+        'is:open is:pr user:googleapis user:GoogleCloudPlatform label:"automerge"',
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:454]: use the constant?

File: packages/merge-on-green/src/merge-on-green.ts [hunk 9]
```diff
@@ -331,6 +432,77 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+    }
+  );
+
+  const issuesAutomergeExactLabel = await github.paginate(
+    await github.search.issuesAndPullRequests,
+    {
+      q:
+        'is:open is:pr user:googleapis user:GoogleCloudPlatform label:"automerge: exact"',
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:462]: use the constant?

File: packages/merge-on-green/src/merge-on-green.ts [hunk 10]
```diff
@@ -331,6 +432,77 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+    if (!pullRequestInDatastore) {
+      logger.info('before adding PR');
+      await handler.addPR(
+        {
+          number: issue.number,
+          owner,
+          repo,
+          state: 'continue',
+          url: issue.html_url,
+          label: 'automerge',
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:477]: use the constant?

File: packages/merge-on-green/src/merge-on-green.ts [hunk 11]
```diff
@@ -331,6 +432,77 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+    const pullRequestInDatastore: WatchPR = await handler.getPR(issue.html_url);
+    if (!pullRequestInDatastore) {
+      await handler.addPR(
+        {
+          number: issue.number,
+          owner,
+          repo,
+          state: 'continue',
+          url: issue.html_url,
+          label: 'automerge: exact',
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:496]: use the constant?

File: packages/merge-on-green/src/merge-on-green.ts [hunk 12]
```diff
@@ -331,6 +432,77 @@ handler.checkPRMergeability = async function checkPRMergeability(
...
+ * @returns void
+ */
+handler.pickUpPR = async function pickUpPR(
+  owner: string,
+  repo: string,
+  github: Context['github']
+) {
+  // Github does not support searching the labels with 'OR'.
+  // The searching for issues is considered to be an "AND" instead of an "OR" .
+  const issuesAutomergeLabel = await github.paginate(
```
Comment (chingor13) [inline packages/merge-on-green/src/merge-on-green.ts:450]: These 2 searches can probably be started at the same time then awaited before we want to start executing on them.

File: packages/merge-on-green/src/merge-on-green.ts (hunk 7 above)
Comment (sofisl) [reply packages/merge-on-green/src/merge-on-green.ts:542]: Oh, I'm expecting this to run on a separate cron job - we won't try merging on the same run as we try picking up the PRs.

**Comment by sofisl** [comment]
@chingor13, really appreciate the in-depth review: it's super helpful to make sure we stay consistent within and outside the package. Thank you!

File: packages/merge-on-green/test/merge-on-green.ts [hunk 13]
```diff
@@ -960,6 +1226,7 @@ describe('merge-on-green', () => {
 
       //This function is supposed to respond with an error
       it('does not add a PR if branch protection errors and comments on PR', async () => {
+        addPRStub.restore();
```
Comment (bcoe) [inline packages/merge-on-green/test/merge-on-green.ts:1229]: one thought here, I think we could use Sinon's sandbox functionality and just always restart the sandbox in a `beforeEach`?

**Review by bcoe — APPROVED** [review]
This is looking good to me, thanks for DRY(don't repeat yourself)ing things up 👍

File: packages/merge-on-green/test/merge-on-green.ts (hunk 13 above)
Comment (sofisl) [reply packages/merge-on-green/test/merge-on-green.ts:1229]: I think we are, line 852.

We need to restore sometimes in order to provide specific functionality for some specific tests.

Comment (chingor13) [reply packages/merge-on-green/test/merge-on-green.ts:1229]: If you need different setup scenarios, you can split describe blocks with separate beforeEach implementations.

**Review by chingor13 — APPROVED** [review]

**Pull request merged**